pytential/linalg/skeletonization.py

```diff
@@ -0,0 +1,790 @@
+        istart, iend = indices._block_ranges[i:i + 2]
+        mat[istart:iend] *= waa[i]
+
+    return mat
+
+
+def _prepare_nearfield_expr(places, expr, auto_where):
+    from pytential.symbolic.execution import _prepare_expr
+    return _prepare_expr(places, expr, auto_where=auto_where)
+
+
+def _prepare_farfield_expr(places, expr, auto_where=None, qbx_forced_limit=None):
+    from pytential.symbolic.mappers import OperatorCollector
+    expr, = OperatorCollector()(expr)
+
+    # ensure all IntGs have the requested forced limit
+    expr = QBXForcedLimitReplacer(qbx_forced_limit=qbx_forced_limit)(expr)
+    # ensure all IntGs have their source and targets set
+    expr = DOFDescriptorReplacer(auto_where[0], auto_where[1])(expr)
+    # ensure all IntGs remove all the kernel derivatives
+    expr = KernelTransformationRemover()(expr)
+
+    return expr
+
+
+@dataclass(frozen=True)
+class SkeletonizationWrangler:
+    """
+    .. attribute:: exprs
+
+        An :class:`~numpy.ndarray` of expressions (layer potentials)
+        that correspond to the output blocks of the matrix.
+
+    .. attribute:: input_exprs
+
+        A :class:`tuple` of densities that correspond to the input blocks
+        of the matrix.
+
+    .. attribute:: domains
+
+        A :class:`tuple` of the same length as *input_exprs* defining the
+        domain of each input.
+
+    .. attribute:: context
+
+        A :class:`dict` with additional parameters required to evaluate the
+        expressions.
+
+    The following attributes and methods are internal and used for skeletonization.
+
+    .. attribute:: weighted_sources
+
+        A flag which if *True* adds a weight to the source to proxy evaluation.
+
+    .. attribute:: weighted_targets
+
+        A flag which if *True* adds a weight to the proxy to target evaluation.
+
+    .. attribute:: source_farfield_block_builder
+    .. attribute:: target_farfield_block_builder
+
+        A callable that is used to evaluate farfield proxy interactions.
+        This should follow the calling convention of the constructor to
+        :class:`pytential.symbolic.matrix.FarFieldBlockBuilder`.
+
+    .. attribute:: nearfield_block_builder
+
+        A callable that is used to evaluate nearfield neighbour interactions.
+        This should follow the calling convention of the constructor to
+        :class:`pytential.symbolic.matrix.NearFieldBlockBuilder`.
+
+    .. automethod:: evaluate_source_farfield
+    .. automethod:: evaluate_target_farfield
+    .. automethod:: evaluate_source_nearfield
+    .. automethod:: evaluate_target_nearfield
+    """
+
+    exprs: np.ndarray
+    input_exprs: Tuple[sym.var, ...]
+    domains: Tuple[Hashable, ...]
+    context: Dict[str, Any]
+
+    # internal
+    weighted_sources: bool
+    weighted_targets: bool
+    source_farfield_block_builder: Callable[..., np.ndarray]
+    target_farfield_block_builder: Callable[..., np.ndarray]
+    nearfield_block_builder: Callable[..., np.ndarray]
```

Comment:
Too many things called "block"?

Reply:
https://github.com/inducer/pytential/pull/142 + renamed all the additional uses in `skeletonization.py` in the same way.